File: cpp_solutions/LinkedList/Swap_Nodes_in_Pairs.cpp
/*
*
* Given a linked list, swap every two adjacent nodes and return its head.

* For example, Given 1->2->3->4, you should return the list as 2->1->4->3.

* Your algorithm should use only constant space. You may not modify the values in the list, only nodes itself can be changed.
*
**/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
		if(!head || !head->next) return head;

		ListNode dummy(0);
		ListNode* p = &dummy;
		dummy.next = head;

		while(p && p->next && p->next->next){
			ListNode* n = p->next;
			ListNode* nn = p->next->next;
			p->next = nn;
			n->next = nn->next;
			nn->next = n;
			p = p->next->next;
		}        

		return dummy.next;
    }
};

File: python_solutions/LinkedList/Swap_Nodes_in_Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        pPrev, pCur = ListNode(0), head
        pPrev.next, dummy= head, pPrev
        
        while pCur:
            if pCur.next is None:
                return dummy.next
            # reverse
            pNext = pCur.next
            pCur.next = pNext.next
            pNext.next = pPrev.next
            pPrev.next = pNext
            
            pPrev = pPrev.next.next
            pCur = pPrev.next
            
        return dummy.next
